halo/evmengine/keeper/abci.go

```diff
@@ -145,7 +147,88 @@ func (k *Keeper) PrepareProposal(ctx sdk.Context, req *abci.RequestPreparePropos
 	return &abci.ResponsePrepareProposal{Txs: [][]byte{tx}}, nil
 }
 
-func submitPayload(ctx context.Context, engineCl ethclient.EngineClient, ts time.Time, proposer common.Address) (engine.ForkChoiceResponse, error) {
+// Finalize is called by our custom ABCI wrapper after a block is finalized.
+// It starts an optimistic build if enabled and if we are the next proposer.
+//
+// This custom ABCI callback is used since we need to trigger optimistic builds
+// immediately after FinalizeBlock with the latest app hash
+// which isn't available from cosmosSDK otherwise.
+func (k *Keeper) Finalize(ctx context.Context, req *abci.RequestFinalizeBlock, resp *abci.ResponseFinalizeBlock) error {
+	if !k.buildOptimistic {
+		return nil // Not enabled.
+	}
+
+	// Maybe start building the next block if we are the next proposer.
+	isNext, err := k.isNextProposer(ctx, req.ProposerAddress, req.Height)
+	if err != nil {
+		return errors.Wrap(err, "next proposer")
+	} else if !isNext {
+		return nil // Nothing to do if we are not next proposer.
+	}
+
+	nextHeight := req.Height + 1
+	ctx = log.WithCtx(ctx, "next_height", nextHeight)
+
+	log.Debug(ctx, "Starting optimistic EVM payload build")
+
+	fcr, err := k.startOptimisticBuild(ctx, common.BytesToHash(resp.AppHash))
+	if err != nil || isUnknown(fcr.PayloadStatus) {
+		log.Warn(ctx, "Starting optimistic build failed", err)
+		return nil
+	} else if isSyncing(fcr.PayloadStatus) {
+		log.Warn(ctx, "Starting optimistic build failed; evm syncing", nil)
+		return nil
+	} else if invalid, err := isInvalid(fcr.PayloadStatus); invalid {
+		log.Error(ctx, "Starting optimistic build failed; invalid payload [BUG]", err)
+		return nil
+	}
+
+	k.setOptimisticPayload(fcr.PayloadID, uint64(nextHeight))
+
+	return nil
+}
+
+func (k *Keeper) startOptimisticBuild(ctx context.Context, appHash common.Hash) (engine.ForkChoiceResponse, error) {
+	latestEBlock, err := k.engineCl.HeaderByNumber(ctx, nil)
+	if err != nil {
+		return engine.ForkChoiceResponse{}, errors.Wrap(err, "get head")
+	}
+
+	ts := uint64(time.Now().Unix())
```

Comment:
If we use `time.Now()` will this be deterministic across all validators or is this executed only by the block proposer?

Reply:
only proposer